core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -621,6 +632,25 @@ private static AnnotatedValueResolver ofRequestObject(String name, AnnotatedElem
                 .build();
     }
 
+    private static AnnotatedValueResolver ofAttribute(String name,
+                                                      AnnotatedElement annotatedElement,
+                                                      AnnotatedElement typeElement, Class<?> type,
+                                                      DescriptionInfo description) {
+        return new Builder(annotatedElement, type, name)
+                .annotationType(Attribute.class)
+                .typeElement(typeElement)
+                .supportDefault(true)
+                .supportContainer(true)
+                .description(description)
+                .resolver(resolverAttribute((ctx, clazz) -> {
+                                                return ctx.context.attr(AttributeKey.valueOf(clazz, name)) != null ?
+                                                       ctx.context.attr(AttributeKey.valueOf(clazz, name)) :
+                                                       ctx.context.attr(AttributeKey.valueOf(name));
```

Comment:
- `AttributeKey.valueOf()` is not free, so we could instantiate it out of the lambda function, so it's calculated only once.
- Instead of using `clazz`, we should use the `prefix` property a user specified in the annotation. Please see my comment on `@Attribute`.

Reply:
Aye, i totally agree. 
As you said, there is no need to get get type by using `clazz`.


```java
        final List<AttributeKey<?>> attrKeys = new ArrayList<>();

        // Order1. User Defined Type (i,e prefix = '...')
        final AttributeKey<?> prefixTypeAttrKey = AttributeKey.valueOf(attr.prefix(), name);
        attrKeys.add(prefixTypeAttrKey);

        // Order2. Declared Type on method parameter (i,e @Attribute int value, in this case int)
        final AttributeKey<?> declaredTypeKey = AttributeKey.valueOf(type, name);
        attrKeys.add(declaredTypeKey);

        // Order3. If Declared Type is primitive, we should register Reference type as well.  (i,e Interger.class)
        if (type.isPrimitive()) {
            final Class<?> mustRefType = Primitives.wrap(type);
            final AttributeKey<?> mustRefTypeKey = AttributeKey.valueOf(mustRefType, name);
            attrKeys.add(mustRefTypeKey);
        }

        // Order4. for AttributeKey.valueOf("noTypeKey")
        final AttributeKey<Object> objectTypeAttrKey = AttributeKey.valueOf(name);
        attrKeys.add(objectTypeAttrKey);
```
- Search for values sequentially in ServiceContext according to the ordering set above.
```java
for (AttributeKey<?> attrKey : attrKeys) {
    final Object value = ctx.context.attr(attrKey);
    if (value != null)
        return value;
}
return null;
```
- and this is search logic. if cost of searching with attr() is quite cheap, we can modify it as stream expression to improve readability.

Reply:
What do you think about making the search logic a little bit more deterministic and unambiguous? i.e.

```
searchlist = []
if annotation has prefix:
    searchlist += AttributeKey.valueOf(prefix, name)
else:
    searchlist += AttributeKey.valueOf(servicetype, name)
    searchlist += AttributeKey.valueOf(name)
```

- We don't need to consider primitive types because service type is never a primitive type.
- It can be confusing when we try Order2 and 4 when `prefix` is specified explicitly.

Reply:
Thanks for your reviews. 🙇‍♂️ 
You are right, i think more deterministic and unambiguous is better! 

I added two search entries for parameter type. 
For example, `void myMethod(@Attribute int val){...}`, in this case parameter type is `int`.
```
searchlist = []
if annotation has prefix:
    searchlist += AttributeKey.valueOf(prefix, name)
else:
    searchlist += AttributeKey.valueOf(servicetype, name)
     // Added
    searchlist += AttributeKey.valueOf(parameterType, name)
    if parameterType is primitive:
      // Added
      searchlist += AttributeKey.valueOf(Primitives.wrap(parameterType), name))
    searchlist += AttributeKey.valueOf(name)
```

Reply:
We should not add the last two cases that uses `parameterType` as the first component of `AttributeKey`. It is not a way we want our users to use this feature.

Reply:
Thanks your kind comments and I'm really sorry for making you spend a lot of time to correct my  misunderstanding. 🙇‍♂️ 
As you said, i fixed it as shown below.  
```
searchlist = []
if annotation has prefix:
    searchlist += AttributeKey.valueOf(prefix, name)
else:
    searchlist += AttributeKey.valueOf(servicetype, name)
    searchlist += AttributeKey.valueOf(name)
```